File: solutions/98 Validate Binary Search Tree/98.cpp
/*
time complexity: O(n)
tags: Tree, Depth-First Search, Binary Search Tree, Binary Tree
note: 當前的節點對左子樹來說必是最大值，當前的節點對右子樹來說必是最小值。更新子樹的上限與下限，
當有子樹小於下限或是大於上限。代表某節點的右樹出現小於該節點的數，或是某節點的左樹出現大於該節點的數
*/

#include <iostream>
#include <limits>
using namespace std;

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    bool recursion(TreeNode *root, long small, long big)
    {
        // base case
        if (root == nullptr)
            return true;
        // divide
        // 當前的節點對左子樹來說必是最大值
        // 當前的節點對右子樹來說必是最小值
        if (recursion(root->left, small, root->val) && recursion(root->right, root->val, big))
        {
            return (root->val < big) && (root->val > small);
        }
        else
            return false;
    }
    
    bool isValidBST(TreeNode *root)
    {
        return recursion(root, LONG_MIN, LONG_MAX);
    }
};

int main()
{
    Solution sol;
    TreeNode node1 = TreeNode(1);
    TreeNode node3 = TreeNode(3);
    TreeNode node6 = TreeNode(6);
    TreeNode node4 = TreeNode(4, &node3, &node6);
    TreeNode node5 = TreeNode(5, &node1, &node4);
    cout << sol.isValidBST(&node4);
}

File: solutions/98 Validate Binary Search Tree/98.py
"""
time complexity: O(n)
tags: Tree, Depth-First Search, Binary Search Tree, Binary Tree
note: 當前的節點對左子樹來說必是最大值，當前的節點對右子樹來說必是最小值。用 DFS 從 left node 開始檢查。
node 的 left child 必須小於該 node，node 的 right child 必須大於該 node，
若 node 的任一子樹回傳為 False ，便向 parent node 回傳 False
"""
# Definition for a binary tree node.
from math import inf
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def helper(node: Optional[TreeNode], low: int, high: int) -> bool:
            if node is None:
                return True
            if node.val >= high or node.val <= low:
                return False
            else:
                # 該 node 的左右子樹的檢查結果都為 True 才向 parent node 回傳 True
                return helper(node.left, low, node.val) and helper(node.right, node.val, high)

        return helper(root, -inf, inf)
